### ISSUE TITLE
--typecheck flag should be configurable in hardhat.config.ts

### ISSUE THREAD

**Issue by startswithaj** [NomicFoundation/hardhat#4015]
### Describe the feature

Recently `typecheck` was added to make typechecking opt in rather than opts out:

https://github.com/NomicFoundation/hardhat/issues/2945

https://github.com/NomicFoundation/hardhat/pull/3100

Is it possible to make this flag configurable in hardhat.config.ts?

**Comment by fvictorio** [comment]
> Is it possible to make this flag configurable in hardhat.config.ts?

Sadly, that's not possible right now. I'll try to explain why.

When you run Hardhat with a typescript configuration, `ts-node` is loaded to run that config. It's at that stage that we have to decide if the code is typechecked or not. So, having a `typecheck` option in the config would mean that we'd need to read the config to decide if we are going to typecheck, but we need to decide if we are going to typecheck before loading the config...

That being said, I think we technically could do something like this:

- If no `--typecheck` flag is passed, load the config without typechecking.
- Before running whatever we are going to run, we check the config and see if the `typecheck` option is set to true.
- If it is, we start over, but with typechecking enabled this time. (Or, alternatively, we just typecheck as a separate step before running the task.)

But that's way easier said than done, and I'm not sure it's worth the effort.

I'm going to keep this issue open to see how much interest there is on this, but to be honest it's unlikely this will be a top priority soon :slightly_frowning_face:

**Comment by startswithaj** [comment]
hey @fvictorio that makes complete sense.

Slightly off topic, but I have this issue where I use the typechain factories for deployments. The bytecode is included in the types/output generated by typechain. This means that if you make a solidity change and don't execute a compile and then rerun hardhat it uses the old bytecode. 

Do you know if there is a way to detect if during a hardhat execution if any files are recompiled? This way I could do what you suggested above exit early and restart hardhat...

**Comment by fvictorio** [comment]
Sorry, I'm not sure I understand. When you run a script or your tests, Hardhat re-compiles the project (unless you explicitly ask it not to). If you are writing a custom task that needs to have the artifacts/typechain types up to date, you can add `await hre.run("compile")` at the beginning.

**Comment by startswithaj** [comment]
> Sorry, I'm not sure I understand. When you run a script or your tests, Hardhat re-compiles the project (unless you explicitly ask it not to). If you are writing a custom task that needs to have the artifacts/typechain types up to date, you can add `await hre.run("compile")` at the beginning.

> When you run a script or your tests, Hardhat re-compiles the project (unless you explicitly ask it not to).

Yep, when you run the tests hardhat recompiles, but the typechain types and factories from the previous compile have already been loaded into memory by ts-node. Because the typechain generated code includes the bytecode for the contracts, when the tests run, it's using the old contract code and deploys old contracts. The next time you run hardhat the new typechain types/factories are loaded into memory and the correct contract bytecode is deployed. 

What I want to be able to do is detect when a recompile has happened when hardhat runs, so I can force and exit and hardhat to be restarted with the newer types/factories.


<img width="1731" alt="Screenshot 2023-06-15 at 2 49 19 pm" src="https://github.com/NomicFoundation/hardhat/assets/3658522/1742d8e5-8690-46be-90eb-21f18af77e1d">

**Comment by fvictorio** [comment]
> Because the typechain generated code includes the bytecode for the contracts, when the tests run, it's using the old contract code and deploys old contracts.

I see your point now. To be honest, I think the best answer is to use `hre.artifacts` instead of the bytecode generated by typechain, but I understand that that might mean a lot of changes for you.

We don't have an easy way to detect if the compile task did compile (or to check if it will). The only thing that comes to mind is (ab)using the `TASK_COMPILE_SOLIDITY_LOG_NOTHING_TO_COMPILE` subtask, doing something like this:

```js
let nothingToCompile = false;
subtask(TASK_COMPILE_SOLIDITY_LOG_NOTHING_TO_COMPILE)
  .setAction(async (args, hre, runSuper) => {
    nothingToCompile = true;

    return runSuper();
  });
```

And then use the `nothingToCompile` flag in your other tasks. It's not a great solution, but it might work for you as a temporary workaround.

**Comment by startswithaj** [comment]
> > Because the typechain generated code includes the bytecode for the contracts, when the tests run, it's using the old contract code and deploys old contracts.
> 
> I see your point now. To be honest, I think the best answer is to use `hre.artifacts` instead of the bytecode generated by typechain, but I understand that that might mean a lot of changes for you.
> 
> We don't have an easy way to detect if the compile task did compile (or to check if it will). The only thing that comes to mind is (ab)using the `TASK_COMPILE_SOLIDITY_LOG_NOTHING_TO_COMPILE` subtask, doing something like this:
> 
> ```js
> let nothingToCompile = false;
> subtask(TASK_COMPILE_SOLIDITY_LOG_NOTHING_TO_COMPILE)
>   .setAction(async (args, hre, runSuper) => {
>     nothingToCompile = true;
> 
>     return runSuper();
>   });
> ```
> 
> And then use the `nothingToCompile` flag in your other tasks. It's not a great solution, but it might work for you as a temporary workaround.

Ahh this is perfect. Thank you @fvictorio. You've saved me a bunch of time.

>  I think the best answer is to use `hre.artifacts`

IMO the best solution is that Typechain doesn't embedded the bytecode in the factories and or only stores the artifact path and then retrieves the bytecode from HRE, or adds the ability to pass in hre.artifacts to the factory. The typechain Factories are really nice to use.

**Comment by startswithaj** [comment]
For anyone that's interested/comes across this issue this is what I ended up with 


```
import {
  TASK_COMPILE_SOLIDITY_LOG_NOTHING_TO_COMPILE,
  TASK_TEST_SETUP_TEST_ENVIRONMENT,
} from 'hardhat/builtin-tasks/task-names';
import { subtask } from 'hardhat/config';
import { DEPLOY_TASK_NAME } from './deployment-scripts/DeploymentTask';

let nothingToCompile = false;
subtask(TASK_COMPILE_SOLIDITY_LOG_NOTHING_TO_COMPILE).setAction(
  async (args, hre, runSuper) => {
    nothingToCompile = true;
    return runSuper();
  },
);

const checkForCompileAndExit: ActionType<TaskArguments> = async (
  args,
  hre,
  runSuper,
) => {
  if (!nothingToCompile) {
    console.log(
      'A Solc recompile was detected. Hardhat needs to be restarted to allow typechain to use the latest bytecode...exiting',
    );
    process.exit(1);
  }

  return runSuper();
};

subtask(TASK_TEST_SETUP_TEST_ENVIRONMENT).setAction(checkForCompileAndExit);
subtask(DEPLOY_TASK_NAME).setAction(checkForCompileAndExit);

```

**Comment by lastperson** [comment]
It seems that Hardhat v3 doesn't offer typechecking in scripts/tests anymore?

**Comment by kanej** [comment]
> It seems that Hardhat v3 doesn't offer typechecking in scripts/tests anymore?

That is correct, Hardhat 3 does not offer typechecking. If you want type checking run tsc directly:

```shell
npx tsc --noEmit
```

I am going to close this issue, we do not intend to make the `--typecheck` flag configurable in Hardhat 2, and it has been removed in Hardhat 3.
